nixos/modules/services/misc/paperless-ng.nix

```
@@ -0,0 +1,251 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+let
+  cfg = config.services.paperless-ng;
+
+  defaultUser = "paperless";
+
+  env = {
+    PAPERLESS_DATA_DIR = cfg.dataDir;
+    PAPERLESS_MEDIA_ROOT = cfg.mediaDir;
+    PAPERLESS_CONSUMPTION_DIR = cfg.consumptionDir;
+    GUNICORN_CMD_ARGS = "--bind=${cfg.address}:${toString cfg.port}";
+  } // cfg.extraConfig;
+
+  manage = let
+    setupEnv = lib.concatStringsSep "\n" (mapAttrsToList (name: val: "export ${name}=\"${val}\"") env);
+  in pkgs.writeShellScript "manage" ''
+    ${setupEnv}
+    exec ${cfg.package}/bin/paperless-ng "$@"
+  '';
+
+  # Secure the services
+  defaultServiceConfig = {
+    BindReadOnlyPaths = [
+      "/nix/store"
+      "-/etc/resolv.conf"
+      "-/etc/nsswitch.conf"
+      "-/etc/hosts"
+      "-/etc/localtime"
+    ];
+    BindPaths = [
+      cfg.consumptionDir
+      cfg.dataDir
+      cfg.mediaDir
+    ];
+    CapabilityBoundingSet = "";
+    # ProtectClock adds DeviceAllow=char-rtc r
+    DeviceAllow = "";
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    NoNewPrivileges = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    # Needs to connect to redis
+    # PrivateNetwork = true;
+    PrivateTmp = true;
+    PrivateUsers = true;
+    ProcSubset = "pid";
+    ProtectClock = true;
+    # Breaks if the home dir of the user is in /home
+    # ProtectHome = true;
```

```suggestion
    ProtectHome = true;
    ReadWritePaths = [
      cfg.consumptionDir
      cfg.dataDir
      cfg.mediaDir
    ];
```

You just have to explicitly allow access if you enable ProtectHome IIRC.

---

I think `ProtectHome` has no benefit at all when used with `TemporaryFileSystem = "/:ro"` and `Bind*Paths`. We should remove the whole `ProtectHome` comment.

---

From the systemd.exec manpage (modified to make this snippet understandable on its own):
> Setting `ProtectHome` to "yes" is mostly equivalent to set `/home/`, `/root`, and `/run/user` in `InaccessiblePaths=`.

`InaccessiblePaths` also hides any mounts that are in subdirectories, so if `dataDir=/home/user/whatever` and `ProtectHome=true`, `dataDir` would not be accessible to this service.

I agrees that `ProtectHome` does not have much of an effect with a temporary root file system. I’m happy to change the comment, but I’d like to keep the line commented out, just to have all the options in the list :)